[2.7][HttpKernel] Added support for registering Bundle dependencies

andrerom opened the pull request:
| Q | A |
| --- | --- |
| Bug fix? | no |
| New feature? | yes |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | #13505 |
| License | MIT |
| Doc PR | https://github.com/symfony/symfony-docs/pull/5103 |

As there are ongoing discussions in both in eZ and [Symfony CMF](https://github.com/symfony-cmf/symfony-cmf/issues/216) on this topic, agreed with @lsmith77 to open PR to discuss this potential light weight native approach. Alternative if native feature is not wanted is [OroPlatform](https://github.com/orocrm/platform/blob/master/src/Oro/Bundle/DistributionBundle/OroKernel.php), a custom, still somewhat project specific and heavier approach involving bundles.yml files and caching needs.
## What?

Adds possibility for a Bundle to specify other bundles it depends on. This allows root/meta projects to not have to specify dependencies of dependencies, and no need for user to perform manual upgrade work in Kernel file when they change.

A dependency implies two things for the Kernel:
- make sure it is loaded, unless dependency is marked as _optional_
- order it before "current" bundle

The last point is because we assume a dependency is something we might extend so we _always_ load it before "self".
## How

A optional interface is added that bundles can implement, while the method could have been added
to the main interface like getParent(), this approach allows the code to check using instanceof,
keeps full BC, and allows a clear contract on use of this feature.

The approach further explicitly avoids giving bundle writers any control over priority of load ordering other
then "before self", as otherwise it would 1. potential lead to conflicts, and 2. complicate the implementation.

For further info on the new function and why it returns FQN class names, see BundleDependenciesInterface.
## Todo
- [x] CS fixes
- [x] Testing, incl recursion testing and fixes in implementation for this
- [x] Doc
## Possible followups
- Could potentially add support for using Ns\MyBundle::CLASS also in registerBundles
- Consider caching the resulting bundle order if needed by additional features
- <del>Open request from Symfony-CMF is to be be able to specify optional dependencies, basically dependencies that should be silently ignored if they don't exists, however as this is during boot that would lead to un cached file system calls (_PHP does not cache stat calls to non existing files_) unless result is cached as mentioned in prior point</del>
- There is some code duplication between ordering logic in parent handling and this, could be removed if getParent over time changes to return FQN as well
- Caching and string representation in registerBundles() can potentially open the door for lazy loading bundles

lavoiesl committed:
[DI][Routing] recursive directory loading

Issue #11045

For now, the Routing DirectoryLoader requires the type `directory`
to be specified so it does not conflict with `AnnotationDirectoryLoader`.
However, this could be refactored.

hason committed:
[WebProfilerBundle] Improved page for logs

Commit:
[HttpKernel] Added support for registering Bundle dependencies

For further info on the new function and why it returns class names see BundleDependenciesInterface.

andrerom commented:
Oh, and as pointed out in other PR's, 2.7 is currently broken (ci and cs).

lsmith77 commented:
One thing to note, with the ability to somewhat control load order it can also mean that people will have less needs to write compiler passes. imho would bring down the barrier for these kinds of things, but I also kind of like that inter-Bundle service configuration is clearly laid out in compiler passes.

As for the need of silent ignoring of optional dependencies, indeed this will add overhead but I would imagine that in many cases it will not really be an issue.

This leads me to an alternative/complementary idea: Simply a command that outputs the code to paste into the Kernel Bundle register method.

andrerom commented:
> This leads me to an alternative/complementary idea: Simply a command that outputs the code to paste into the Kernel Bundle register method.

could somehow work, but that would essentially be like a cache the user needs to manually maintain then. And won't be able to execute the command if bundle dependencies goes away after an composer update :/

lolautruche commented on src/Symfony/Component/HttpKernel/Kernel.php:
```diff
@@ -506,6 +507,89 @@ protected function initializeBundles()
+                $dependency = $topMostBundlesMap[$dependencyFQN];
+            } else {
+                if (!class_exists($dependencyFQN)) {
+                    throw new \RuntimeException(sprintf("Could not find Bundle '%s', set as dependency by '%s'", $dependencyFQN, $parent));
```
I'd go for a specific exception here, for clarity sake.

andrerom replied:
yes, where looking for better examples, open for suggestions.

KingCrunch replied:
`DependencyMismatchException extends \RuntimeException`? Just to throw something in

andrerom replied:
Thanks, but something like `DependencyNotFoundException` would be even more accurate right? or did you have something in mind with the name?

KingCrunch replied:
Just thought, that maybe you can reuse it for other dependency issues as well. On the other hand it's just a suggestion :smile:

hason replied:
What if the bundle is optional dependency?

andrerom replied:
that is currently not supported, see comments above. But if it was, it should of course not throw ;)

andrerom replied:
Found `ClassNotFoundException`, seems to be the most fitting here.
@KingCrunch What do you think?

andrerom replied:
Or not, it has hard requirement on previous exception.

lolautruche commented on src/Symfony/Component/HttpKernel/Kernel.php:
```diff
@@ -506,6 +507,89 @@ protected function initializeBundles()
+                if (!class_exists($dependencyFQN)) {
+                    throw new \RuntimeException(sprintf("Could not find Bundle '%s', set as dependency by '%s'", $dependencyFQN, $parent));
+                }
+                $dependency = new $dependencyFQN();
```
What if the bundle needs arguments? Simple example is [`EzPublishLegacyBundle`](https://github.com/ezsystems/LegacyBridge/blob/master/bundle/EzPublishLegacyBundle.php#L29)

andrerom replied:
Not supported here as mentioned in BundleDependenciesInterface doc.

mvrhov commented:
Hm.. On one side we are removing magic (twig automatic file handling) on the other we are adding it. Is it really that hard to write one file into the AppKernel. After all you also have to provide configuration for that bundle

andrerom commented:
Good point, but you would only need that if defaults are not ok, right? (side: didn't fully get what twig file handling is in this regard btw)

Also in this scenario where root1 depends on dependency2, it might be that root1 makes sure to add own events/tags that are used by dependency2.  Root1 might even adjust the defaults of dependency2 to better fit the combined use for these two bundles. Meaning defaults can often be ok.

With the goal of simplifying distribution handling and upgrade needs, do you see other features needed around this? For instance on configuration?

mmoreram commented:
Thanks GOD for this PR...

andrerom commented:
Tests and further improvements done, doc in review, so ready for second wave of review now :tada:

hhamon commented on src/Symfony/Component/HttpKernel/Exception/DependencyMismatchException.php:
```diff
@@ -0,0 +1,41 @@
+
+namespace Symfony\Component\HttpKernel\Exception;
+
+use RuntimeException;
```
Should be removed in favor of an inline `new \RuntimeException()` in the code.

andrerom replied:
Hi, you mean removing DependencyMismatchException?
That was the original case before this review thread:
https://github.com/symfony/symfony/pull/13945#discussion-diff-26562986

or did you imply inline `extends \RuntimeException`

hhamon replied:
Good ;)

hhamon commented on src/Symfony/Component/HttpKernel/Kernel.php:
```diff
@@ -506,6 +508,95 @@ protected function initializeBundles()
+     *
+     * @uses appendDependenciesRecursively
+     *
+     * @return \Symfony\Component\HttpKernel\Bundle\BundleInterface[] An array of bundle instances.
```
`@return BundleInterface[]` with the `use` statement at the top if not already set.

lsmith77 commented:
it might be a good idea to also offer some debugging tool to help developers understand the dependency tree. though this is imho rather a nice to have that could come via a separate PR.

lsmith77 commented on src/Symfony/Component/HttpKernel/Tests/Fixtures/BundleDependencies/BundleCDependenciesBA.php:
```diff
@@ -0,0 +1,25 @@
+    public function getBundleDependencies()
+    {
+        return array(
+            'Symfony\Component\HttpKernel\Tests\Fixtures\BundleDependencies\BundleBDependenciesA',
```
if we are to ever add optional support imho it would be better to have an associative array here, where the value would be true for required and false for optional for example. then when doing the class_exists check you would simply not throw an exception for optional dependencies.

andrerom replied:
Ok, thats a fair point, as we should decide on the format now and not change it later. I was thinking separate method, maybe even on a separate `BundleOptionalDependenciesInterface` if this would be done in a followup.

andrerom replied:
What about
```php
class AcmeBundle extends Bundle implements BundleDependenciesInterface
{
    public function getBundleDependencies()
    {
        return array(
            self::REQUIRED => array(
                // All values must be FQN strings to avoid bundles being loaded several times
                'FOS\HttpCacheBundle\FOSHttpCacheBundle',
                // if on php 5.5 and higher
                Oneup\FlysystemBundle\OneupFlysystemBundle::CLASS,
            ),
            self::OPTIONAL => array(
                // Optional dependencies can not use ::CLASS constant
                'FOS\HttpCacheBundle\FOSHttpCacheBundle',
        );
    }
}
 ```

lsmith77 replied:
sure .. but what would be the advantage? or are you just proposing a possible new format with migration path in case we leave it out today and want to add optional deps later?

just seems like a more complicated structure. plus an assoc array implicitly takes care of duplicates. then again if one ends up having so many dependencies that accidental duplicates are a risk, one might be over using the feature :)

andrerom replied:
> or are you just proposing a possible new format with migration path in case we leave it out today and want to add optional deps later?

y

andrerom replied:
returning to the specifics:

> just seems like a more complicated structure. plus an assoc array implicitly takes care of duplicates. then again if one ends up having so many dependencies that accidental duplicates are a risk, one might be over using the feature :)

Was also thinking less typing, implementation can not do a simple array join anyway with your proposed structure, as you always would want REQUIRED to overwrite OPTIONAL. However I'll try out your structure to see here now.

andrerom commented:
> it might be a good idea to also offer some debugging tool to help developers understand the dependency tree. 

So like `$this->bundleMap` for parent/child relationships then, could potentially have information about both in the same structure.

andrerom commented:
So quick poll: 
| Q             | A
| ------------- | ---
| Add optional dependencies now?      | yes/no

*Context: Despite no cache for now, so it will be doing a uncached class_exists -> file_exists lookup for missing dependencies for now*

EmmanuelVella commented on src/Symfony/Component/HttpKernel/Bundle/BundleDependenciesInterface.php:
```diff
@@ -0,0 +1,64 @@
+     *             'FOS\HttpCacheBundle\FOSHttpCacheBundle',
+     *
+     *             // If you require PHP 5.5 or higher it is better to use `::CLASS` constant:
+     *             Oneup\FlysystemBundle\OneupFlysystemBundle::CLASS
```
I think you should use lower case for ``class`` : http://php.net/manual/en/language.oop5.basic.php#language.oop5.basic.class.class

andrerom replied:
Ok, tried to look for examples and those I found (probably outside of php doc) where using uppercase. fixed now :)

mmoreram commented:
Hey guys. We started developing this feature some days ago as well (I did not realized about this PR till yesterday).

https://github.com/elcodi/elcodi/issues/639

You can take a look to our implementation. Maybe is helpful :)

andrerom commented:
> You can take a look to our implementation. Maybe is helpful :)

tried, but it is a lot of code, what I managed to see is that it does not take care about ordering.

andrerom commented:
Core dev review ping: @stof @weaverryan (also see doc pr)

lsmith77 commented:
ping @symfony/deciders

lsmith77 commented:
I am +1 on adding optional support as its basically just about not throwing an exception

andrerom commented:
ok

Commit:
[HttpKernel] Added support for registering *optional* Bundle dependencies

andrerom commented:
added *(as second commit)*, review ping @lsmith77

Pierstoval commented:
@andrerom Setting bundle dependencies this way is good but I find it a bit more complex than #13506 :confused:

andrerom commented:
@Pierstoval damn, overlooked that there was another PR.

As for format:
- See first commit, it was simpler, discussion with @lsmith77 around current structure can be found [here](https://github.com/symfony/symfony/pull/13945#discussion_r27099451)
- Reason why I didn't went for instances was that you will get conflicting instances, so if someone adds state to the bundle via constructor that might get different result then they expect. On the other side this is such an edge case it could probably just be documented to keep the use simple.

Pierstoval commented:
The 2 differences I see with my first implementation (I almost abandoned the PR when I saw only one positive feedback), is that there is no BC because you use a `BundleDependenciesInterface`, and that you allow optional bundles to be used.
 
Sounds great to me, can't wait to see it implemented as I really would like to use these bundle dependencies :+1:

Commit:
[HttpKernel] Simplify Bundle dependencies by moving *optional* to doc

fabpot committed:
Revert "removed 3.0 constraints from 2.7 composer files"

This reverts commit 5503804cf91f4ede81b94b34609653c013139038.

fabpot committed:
updated version to 2.8

fabpot committed:
Merge branch '2.7' into 2.8

* 2.7:
  [HttpKernel] fixed a regression when no exception listeners are registered
  renamed some confusing tests
  propel/propel1 is now useless
  bumped Symfony version to 2.7.0
  updated VERSION for 2.7.0-BETA1
  updated CHANGELOG for 2.7.0-BETA1
  [2.3] Fix @link annotations
  [2.7] For @link annotations
  [2.6] Fix @link annotations
  [2.7][Console] Count the  array instead of
  [Config][cache factory] check type of callback argument.
  Fix javascript
  [2.3][Translation] test refresh cache when resources File change.
  Added deprecation message and original color back
  [Translator] Cache does not take fallback locales into consideration

sgehrig committed:
added the --all option to the debug:translation command

derrabus committed:
Support for array denormalization.

aitboudad committed:
[Translation] added FileLoader.

aitboudad committed:
[FrameworkBundle][Server Command] add address port number option.

fabpot committed:
feature #14383 [FrameworkBundle][Server Command] add address port number option. (aitboudad)

This PR was merged into the 2.8 branch.

Discussion
----------

[FrameworkBundle][Server Command] add address port number option.

| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | ~
| License       | MIT
| Doc PR        | ~

This allow running server without repeating `127.0.0.1` each time, in the most of case we only change the port number.

* Before
```
app/console server:start 127.0.0.1:8080
```

* After
```
app/console server:start -p 8080
```

Commits
-------

fbe1a43 [FrameworkBundle][Server Command] add address port number option.

fabpot committed:
feature #14359 [Translation] added FileLoader. (aitboudad)

This PR was merged into the 2.8 branch.

Discussion
----------

[Translation] added FileLoader.

| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | ~
| License       | MIT
| Doc PR        | ~

Commits
-------

3694e5e [Translation] added FileLoader.

Tobion committed:
[Form] deprecate read_only option

lyrixx committed:
[VarDumper] Added support for SplFileInfo

lyrixx committed:
[VarDumper] Added support for SplFileObject

lyrixx committed:
[Console] Bind the closure (code) to the Command if possible

gepo committed:
[Translation] add options 'as_tree', 'inline' to YamlFileDumper to dump messages as tree instead of simple list.

Dump messages as tree based on '.' character as a delimeter in path.
For example this rray('foo.bar' => 'value') will be converted to array('foo' => array('bar' => 'value')).

Correctly process cases like this ['foo.bar' => 'test1', 'foo' => 'test2'].

nicolas-grekas committed:
feature #14424 [VarDumper] Added support for SplFileInfo (lyrixx)

This PR was merged into the 2.8 branch.

Discussion
----------

[VarDumper] Added support for SplFileInfo

| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | ~
| License       | MIT
| Doc PR        | ~

Before:
```
SplFileInfo {#3}
```

After:
```
SplFileInfo {#3
  path: "/home/greg/dev/github.com/lyrixx/gaga-photo/fixtures"
  filename: "grisou.jpg"
  basename: "grisou.jpg"
  pathname: "/home/greg/dev/github.com/lyrixx/gaga-photo/fixtures/grisou.jpg"
  extension: "jpg"
  realPath: "/home/greg/dev/github.com/lyrixx/gaga-photo/fixtures/grisou.jpg"
  aTime: 2015-04-20 23:21:15
  mTime: 2014-08-17 15:37:36
  cTime: 2015-04-20 23:09:51
  inode: 12197643
  size: 966553
  perms: 0100750
  owner: 1000
  group: 1000
  type: "file"
  writable: true
  readable: true
  executable: true
  file: true
  dir: false
  link: false
}
```

For a link:

```
SplFileInfo {#3
  path: ""
  filename: "link"
  basename: "link"
  pathname: "link"
  extension: ""
  realPath: "/home/greg/dev/github.com/symfony/var-dumper/composer.json"
  aTime: 2015-04-20 23:39:15
  mTime: 2015-04-20 23:39:15
  cTime: 2015-04-20 23:39:15
  inode: 11272232
  size: 967
  perms: 0100644
  owner: 1000
  group: 1000
  type: "link"
  writable: true
  readable: true
  executable: false
  file: true
  dir: false
  link: true
  linkTarget: "composer.json"
}
```

For a remote file:

```
SplFileInfo {#3
  path: "http://google.com"
  filename: "foobar.php"
  basename: "foobar.php"
  pathname: "http://google.com/foobar.php"
  extension: "php"
  realPath: false
  writable: false
  readable: false
  executable: false
  file: false
  dir: false
  link: false
}
```

For a CSV:
```
SplFileObject {#3
  path: ""
  filename: "test.csv"
  basename: "test.csv"
  pathname: "test.csv"
  extension: "csv"
  realPath: "/home/greg/dev/github.com/symfony/var-dumper/test.csv"
  aTime: 2015-04-21 12:55:07
  mTime: 2015-04-21 12:55:07
  cTime: 2015-04-21 12:55:07
  inode: 11272237
  size: 0
  perms: 0100644
  owner: 1000
  group: 1000
  type: "file"
  writable: true
  readable: true
  executable: false
  file: true
  dir: false
  link: false
  csvControl: array:2 [
    0 => ","
    1 => """
  ]
  flags: SKIP_EMPTY|READ_CSV
  maxLineLen: 0
  fstat: array:7 [
    "dev" => 64513
    "ino" => 11272237
    "nlink" => 1
    "rdev" => 0
    "blksize" => 4096
    "blocks" => 0
    "…" => "…20"
  ]
  eof: false
  key: 0
}

Commits
-------

e75c233 [VarDumper] Added support for SplFileObject
eb50b13 [VarDumper] Added support for SplFileInfo

nicolas-grekas committed:
[VarDumper] Allow preserving a subset of cut arrays

stof committed:
feature #14431 [Console] Bind the closure (code) to the Command if possible (lyrixx)

This PR was merged into the 2.8 branch.

Discussion
----------

[Console] Bind the closure (code) to the Command if possible

| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | ~
| License       | MIT
| Doc PR        | ~

This allow this kind of code:

```php
#!/usr/bin/env php
<?php

require __DIR__.'/vendor/autoload.php';

use Symfony\Component\Console\Application;
use Symfony\Component\Console\Command\Command;
use Symfony\Component\Console\Input\InputInterface;
use Symfony\Component\Console\Output\OutputInterface;

$application = new Application();

$application->add((new Command('process')))
    ->setDescription('Play all other commands')
    ->setCode(function (InputInterface $input, OutputInterface $output) use ($application) {
        $application = $this->getApplication();
        $help = $application->find('help');
        $output->writeln($help->getHelp());
    })
;
$application->run();
```

Commits
-------

ff4424a [Console] Bind the closure (code) to the Command if possible

nicolas-grekas committed:
feature #14443 [VarDumper] Allow preserving a subset of cut arrays (nicolas-grekas)

This PR was merged into the 2.8 branch.

Discussion
----------

[VarDumper] Allow preserving a subset of cut arrays

| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | -
| License       | MIT
| Doc PR        | -

Commits
-------

208ca94 [VarDumper] Allow preserving a subset of cut arrays

nicolas-grekas committed:
Merge branch '2.7' into 2.8

* 2.7: (40 commits)
  [Debug] Fix ClassNotFoundFatalErrorHandler candidates lookups
  [2.6][Translator] Extend, refactor and simplify Translator tests.
  Update DebugClassLoader.php
  inject asset packages in assets helper service
  [travis] Do not exclude legacy tests on 2.7
  [HttpFoundation] remove getExtension method
  [2.6][Translation] fix legacy tests.
  [Form] Removed remaining deprecation notices in the test suite
  [Form] Moved deprecation notice triggers to file level
  [Debug] Map PHP errors to LogLevel::CRITICAL
  [Routing][DependencyInjection] Support .yaml extension in YAML loaders
  [DX] improve file loader error for router/other resources in bundle
  [FrameworkBundle] Initialize translator with the default locale.
  [FrameworkBundle] Fix Routing\DelegatingLoader resiliency to fatal errors
  [2.7][Translation] remove duplicate code for loading catalogue.
  [2.6][Translation] remove duplicate code for loading catalogue.
  [HttpKernel] Cleanup ExceptionListener
  CS fixes
  [DependencyInjection] Show better error when the Yaml component is not installed
  [2.3] SCA for Components - reference mismatches
  ...

nicolas-grekas committed:
Merge branch '2.7' into 2.8

* 2.7:
  [Debug] Fix deprecated use of DebugClassLoader
  Revert "[HttpKernel] Throw a LogicException when kernel.exception does not led to a Response"
  [Validator] Fixed Choice when an empty array is used in the "choices" option

nicolas-grekas committed:
Merge branch '2.7' into 2.8

* 2.7:
  Fix test

nicolas-grekas committed:
Merge branch '2.7' into 2.8

* 2.7:
  [FrameworkBundle] Fix test related to PSR-4

Commit:
Changed output for translation:debug command

aitboudad committed:
feature #14473 [DX] Minor improvement for the translation:debug output (nicolasdewez)

This PR was submitted for the master branch but it was merged into the 2.8 branch instead (closes #14473).

Discussion
----------

[DX] Minor improvement for the translation:debug output

| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #13874

Commits
-------

b941bd3 Changed output for translation:debug command

aramalipoor commented:
Why not add an `options` ability so that a lot of things like restricting on environments, kernel injection, optional flag, etc. would be possible?

```php
return [
    Acme\Bundle\BlogBundle\AcmeBlogBundle::class,
    Acme\Bundle\AnotherBundle\AcmeAnotherBundle::class => ['inject_kernel' => true],
    Acme\Bundle\ThirdBundle\AcmeThirdBundle::class => ['optional' => true],
    Acme\Bundle\DebugBundle\AcmeDebugBundle::class => ['environments' => ['dev', 'test']],
];
````

aramalipoor commented on src/Symfony/Component/HttpKernel/Bundle/BundleDependenciesInterface.php:
```diff
@@ -0,0 +1,69 @@
+     *
+     * @api
+     */
+    public function getBundleDependencies();
```
We can pass $kernel to this method so the developer can e.g check for environments.
```php
public function getBundleDependencies(KernelInterface $kernel);
```

Pierstoval commented:
@aramalipoor's proposal is great: having a bundles list and a "variable" key=>value pair allows more flexibility for both simple and complex dependencies. 
 
Where are "we" about this feature? Can't wait to see it in the first SF3 beta :smile:

andrerom commented:
everything is possible, but from my pov I'm trying to look into things to improve on performance for boot section, so that means typically limiting what kind of logic is exposed to be able to do in these steps, so things like dependency and parent/child traversing can be cached.

I'm not so big a fan of adding options like that, but the examples you provided are worth tackling separately:
'optional' is proposed in two forms here, see how ae35ae0 propose changing it from being a feature complicating the format, to one simply covered by doc.
'environments' is good input, can and should be injected, cache would anyway very by this
'inject_kernel': this pr kind of goes in the direction of what the doc already says on not relying to much on logic in ctor, but could just as well do it as a convention, or allow bundles to implement `KernelAware` :)

aitboudad committed:
feature #14320 [Translation] added an --all option to the debug:translation command #14237 (sgehrig)

This PR was submitted for the 2.7 branch but it was merged into the 2.8 branch instead (closes #14320).

Discussion
----------

[Translation] added an --all option to the debug:translation command #14237

| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #14237
| License       | MIT
| Doc PR        | -

- [ ] submit changes to the documentation

This PR adds an `--all` option the `debug:translation` command to enable debugging translations in all registered bundles *AND* the app directory.

Commits
-------

0e7fe79 added the --all option to the debug:translation command

aitboudad committed:
[Translator] deprecate getMessages in favor of getCatalogue.

Seldaek committed:
[FrameworkBundle][DX] Add option to specify additional translation loading paths

Seldaek committed:
[FrameworkBundle][EventDispatcher] Add priorities to the debug:event-dispatcher command

aitboudad committed:
feature #14546 [Translator] deprecate getMessages in favor of getCatalogue. (aitboudad)

This PR was merged into the 2.8 branch.

Discussion
----------

[Translator] deprecate getMessages in favor of getCatalogue.

| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Fixed tickets  | #14529
| Tests pass?   | yes
| License       | MIT

Commits
-------

2869a32 [Translator] deprecate getMessages in favor of getCatalogue.

fabpot committed:
feature #14563 [FrameworkBundle][EventDispatcher] Add priorities to the debug:event-dispatcher command (Seldaek)

This PR was merged into the 2.8 branch.

Discussion
----------

[FrameworkBundle][EventDispatcher] Add priorities to the debug:event-dispatcher command

| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets |
| License       | MIT
| Doc PR        | TODO

I find it helps to figure out which priority is needed to override another event listener without digging in the sources.

Commits
-------

952929c [FrameworkBundle][EventDispatcher] Add priorities to the debug:event-dispatcher command

fabpot committed:
fixed CS

aitboudad committed:
feature #14561 [FrameworkBundle][DX] Add option to specify additional translation loading paths (Seldaek)

This PR was squashed before being merged into the 2.8 branch (closes #14561).

Discussion
----------

[FrameworkBundle][DX] Add option to specify additional translation loading paths

| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #971
| License       | MIT
| Doc PR        | TODO

Thanks to this PR, next time I want to just do:

```
    translator:
        fallback: "%locale%"
        paths: ['%kernel.root_dir%/../vendor/internal/package/translations']
```

I will be able to, and won't have to dig in docs, start building a translation loader before I realize it's not what I want, dig in the framework ext, and then have to write a compiler pass to inject my file to the translator resource paths.

Commits
-------

bba0a25 [FrameworkBundle][DX] Add option to specify additional translation loading paths

nicolas-grekas committed:
Merge branch '2.7' into 2.8

* 2.7: (70 commits)
  [travis] Use container-based infrastructure
  [HttpKernel] use ConfigCache::getPath() method when it exists
  [PropertyAccess] Fix setting public property on a class having a magic getter
  [Routing] Display file which contain deprecated option
  ContainerInterface: unused exception dropped
  bumped Symfony version to 2.6.8
  updated VERSION for 2.6.7
  updated CHANGELOG for 2.6.7
  bumped Symfony version to 2.3.29
  updated VERSION for 2.3.28
  update CONTRIBUTORS for 2.3.28
  updated CHANGELOG for 2.3.28
  [Debug] Fixed ClassNotFoundFatalErrorHandlerTest
  [SecurityBundle] use access decision constants in config
  [SecurityBundle] use session auth constants in config
  PhpDoc fix in AbstractRememberMeServices
  [Filesystem] Simplified an if statement
  [SecurityBundle] Use Enum Nodes Instead Of Scalar
  [Debug 2.3] Fix test for PHP7
  [HttpKernel] Check if "symfony/proxy-manager-bridge" package is installed
  ...

Conflicts:
	src/Symfony/Bundle/DebugBundle/composer.json
	src/Symfony/Bundle/FrameworkBundle/Command/ServerRunCommand.php
	src/Symfony/Bundle/FrameworkBundle/Command/TranslationDebugCommand.php
	src/Symfony/Component/Form/README.md
	src/Symfony/Component/Intl/README.md
	src/Symfony/Component/Security/README.md
	src/Symfony/Component/Translation/Loader/CsvFileLoader.php
	src/Symfony/Component/Translation/Loader/IniFileLoader.php
	src/Symfony/Component/Translation/Loader/MoFileLoader.php
	src/Symfony/Component/Translation/Loader/PhpFileLoader.php
	src/Symfony/Component/Translation/Loader/PoFileLoader.php
	src/Symfony/Component/Translation/Loader/YamlFileLoader.php
	src/Symfony/Component/Translation/README.md
	src/Symfony/Component/Translation/Translator.php
	src/Symfony/Component/Validator/README.md

nicolas-grekas committed:
Merge branch '2.7' into 2.8

* 2.7:
  [TranslationDebugCommand] fixed failing tests.

nicolas-grekas committed:
Merge branch '2.7' into 2.8

* 2.7:
  fix missing links to https://symfony.com
  fix missing links to https://symfony.com
  [travis] Don't use the cache
  [DebugBundle] Remove inlined dumps on XHR

Conflicts:
	src/Symfony/Component/Security/Acl/README.md
	src/Symfony/Component/Security/Core/README.md
	src/Symfony/Component/Security/Csrf/README.md
	src/Symfony/Component/Security/Http/README.md

nicolas-grekas committed:
Merge branch '2.7' into 2.8

* 2.7:
  [EventDispatcher] make listeners removable from an executed listener
  [2.7][DoctrineBridge] change Logger interface in tests to use Psr\\Log\\LoggerInterface instead of Symfony\\Component\\HttpKernel\\Log\\LoggerInterface"
  [HttpFoundation] Fix volatile MongoDbSessionHandlerTest::testRead()
  bumped Symfony version to 2.7.0
  updated VERSION for 2.7.0-BETA2
  updated CHANGELOG for 2.7.0-BETA2
  Fixed compatibility with PHP7 and up by introducing new constraints (IsNull, IsTrue, IsFalse) and related validators (IsNullValidator, IsTrueValidator, IsFalseValidator)
  [Console] SymfonyStyle: fix block rpadding when escaping '<'
  [Console] fix code style

Conflicts:
	src/Symfony/Component/Form/composer.json

fabpot committed:
Merge branch '2.7' into 2.8

* 2.7:
  fixed CS
  fixed CS
  fixed CS
  Fix WebProfilerBundle compatiblity with HttpKernel < 2.7
  [Validator] Deprecated PHP7-incompatible constraints and related validators
  [DebugBundle] Allow alternative destination for dumps
  [DebugBundle] Use output mechanism of dumpers instead of echoing
  [DebugBundle] Always collect dumps
  [FrameworkBundle] Applied new styles to the config:debug & config:dump-reference commands
  Fix tests in HHVM
  CS: Pre incrementation/decrementation should be used if possible

Conflicts:
	src/Symfony/Bundle/FrameworkBundle/composer.json

fabpot committed:
feature #13324 [WebProfilerBundle] Improved page for logs (hason)

This PR was submitted for the 2.7 branch but it was merged into the 2.8 branch instead (closes #13324).

Discussion
----------

[WebProfilerBundle] Improved page for logs

| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | -
| License       | MIT
| Doc PR        | -

*Before*
![logs-before](https://cloud.githubusercontent.com/assets/288535/5661713/3942c530-972f-11e4-89cb-2e185a6d07ac.png)
*After*
![logs-after](https://cloud.githubusercontent.com/assets/288535/5661681/d02b50da-972e-11e4-96e8-0572428ba82f.png)

Commits
-------

5fdc650 [WebProfilerBundle] Improved page for logs

nicolas-grekas committed:
Added CHANGELOG entries, cleanups

fabpot committed:
feature #14403 [Form] deprecate read_only option (Tobion)

This PR was squashed before being merged into the 2.8 branch (closes #14403).

Discussion
----------

[Form] deprecate read_only option

| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | no
| BC breaks?    | no
| Deprecations? | yes
| Tests pass?   | yes
| Fixed tickets | #10658
| License       | MIT
| Doc PR        | https://github.com/symfony/symfony-docs/pull/3782

Replaces #10676 with a slightly different implementation.
- fixes BC break when 'read_only' => true option and at the same time custom attributes are used by using a normalizer
- adds deprecation notice
- keeps legacy tests

Commits
-------

53330e1 [Form] deprecate read_only option

aitboudad committed:
feature #14630 [Translator] Dump translation constants as tree instead of simple list (gepo)

This PR was merged into the 2.8 branch.

Discussion
----------

[Translator] Dump translation constants as tree instead of simple list

| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | no
| Fixed tickets |
| License       | MIT
| Doc PR        |

PR #14434 for 2.8 branch

Commits
-------

29ec5ca [Translation] add options 'as_tree', 'inline' to YamlFileDumper to dump messages as tree instead of simple list.

nicolas-grekas committed:
[Security] Add setVoters() on AccessDecisionManager

dunglas committed:
feature #14756 [Serializer] Support for array denormalization (derrabus)

This PR was merged into the 2.8 branch.

Discussion
----------

[Serializer] Support for array denormalization

| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | none
| License       | MIT
| Doc PR        | none (yet)

This is a rebase of #14343 against the 2.8 branch.

I had to implement a couple of API endpoints that receive data as JSON-serialized arrays of objects. The current implementation of the Serializer component is capable of serializing such arrays, but so far this operation is a one-way-road. This PR is an attempt to change this.

## Demo
```php
class Dummy
{
    public $foo;
    public $bar;

    public function __construct($foo = null, $bar = null)
    {
        $this->foo = $foo;
        $this->bar = $bar;
    }
}

$serializer = new \Symfony\Component\Serializer\Serializer(
    array(
        new \Symfony\Component\Serializer\Normalizer\PropertyNormalizer(),
        new \Symfony\Component\Serializer\Normalizer\ArrayDenormalizer()
    ),
    array(
        new \Symfony\Component\Serializer\Encoder\JsonEncoder()
    )
);

$json = $serializer->serialize(
    array(
        new Dummy('one', 'two'),
        new Dummy('three', 'four')
    ),
    'json'
);

echo $json . "\n\n";

// Deserialize the JSON string, so we get back to where we started from.
$data = $serializer->deserialize($json, 'Dummy[]', 'json');

var_dump($data);
```

By appending `[]` to the type parameter, you indicate that you expect to deserialize an array of objects of the given type. This is the same notation that phpDocumentor uses to indicate collections. The denormalization of the array is implemented recursively: The denormalizer simply calls `Serializer::denormalize()` on each element of the array. This way, the ArrayDenormalizer can be combined with any other denormalizer.

## Side effects

For this implementation, I had to touch `GetSetMethodNormalizer`, `PropertyNormalizer` and `CustomNormalizer`. Those classes expected `supportsDenormalization` to be called with a valid class in the `$type` parameter. Instead of throwing a reflection exception, they now simply return false. I'm not exactly sure, if this is should be considered to be a BC break.

This implementation violates the `SerializerInterface` which declared `deserialize()` to always return an object. But imho, the assumption that serialized data always represents an object is too restrictive anyway. Also, this declaration is not consistent with the `serialize()` method which accepts `mixed` as input type for the data to serialize.

## Other PRs

I've found an older PR adressing this issue, #12066. That PR was closed because the contributor did not reply to feedback.

Commits
-------

0573f28d Support for array denormalization.

chbruyand committed:
[TwigBundle] Reconfigure twig paths when they are updated

fabpot committed:
feature #14733 [Security] Add setVoters() on AccessDecisionManager (nicolas-grekas)

This PR was merged into the 2.8 branch.

Discussion
----------

[Security] Add setVoters() on AccessDecisionManager

| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | partially #12360
| License       | MIT
| Doc PR        | -

Alternative for #14550

Commits
-------

3fd7cea [Security] Add setVoters() on AccessDecisionManager

fabpot committed:
Merge branch '2.7' into 2.8

* 2.7: (95 commits)
  [DependencyInjection] provide better error message when using deprecated configuration options
  [console][TableCell] get cell width without decoration.
  Improve the config validation in TwigBundle
  [VarDumper] Changed tooltip to expand-all keybinding in OS X
  [Bridge\PhpUnit] Fix composer installed phpunit detection
  [VarDumper] Fix generic casters calling order
  [2.7][SecurityBundle] Remove SecurityContext from Compile
  [WebProfilerBundle][logger] added missing deprecation message.
  Fix profiler CSS
  [Security][Acl] enforce string identifiers
  [FrameworkBundle] make `templating.helper.router` service available again for BC reasons
  [BrowserKit] Fix bug when uri starts with http.
  bumped Symfony version to 2.7.1
  updated VERSION for 2.7.0
  updated CHANGELOG for 2.7.0
  bumped Symfony version to 2.6.10
  updated VERSION for 2.6.9
  updated CHANGELOG for 2.6.9
  fixed tests
  bumped Symfony version to 2.3.31
  ...

Conflicts:
	src/Symfony/Bundle/FrameworkBundle/Command/TranslationDebugCommand.php
	src/Symfony/Bundle/WebProfilerBundle/Resources/views/Collector/logger.html.twig
	src/Symfony/Component/HttpKernel/Kernel.php
	src/Symfony/Component/Translation/Loader/JsonFileLoader.php

fabpot committed:
feature #14700 [DependencyInjection] [Routing] [Config] Recursive directory loading (lavoiesl, nicolas-grekas)

This PR was merged into the 2.8 branch.

Discussion
----------

[DependencyInjection] [Routing] [Config] Recursive directory loading

| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #13246, #11045, #11059
| License       | MIT
| Doc PR        | -

Commits
-------

60b1c5b Added CHANGELOG entries, cleanups
73f0ee2 [DI][Routing] recursive directory loading

fabpot committed:
Merge branch '2.7' into 2.8

* 2.7:
  [Security] Update tests after a merge
  [Console] Remove an unused argument and fix a small cs issue
  [Translator] avoid serialize unserializable resources.

fabpot committed:
feature #14781 [TwigBundle] Reconfigure twig paths when they are updated (chbruyand)

This PR was squashed before being merged into the 2.8 branch (closes #14781).

Discussion
----------

[TwigBundle] Reconfigure twig paths when they are updated

| Q             | A
| ------------- | ---
| Bug fix?      | yes
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #14771, #14768, #14262, #14778
| License       | MIT

Refresh twig paths upon creation and deletion. As we don't care neither about path's modification time nor path's content, a new Resource has been added in the Config Component.
Full discussion in #14778.

Commits
-------

3cbff05 [TwigBundle] Reconfigure twig paths when they are updated

jbgomond committed:
Removed route name from the debugbar.

jbgomond committed:
[toolbar] Merged colored icons in toolbar

OskarStark committed:
added missing "

fabpot committed:
minor #14909 added missing " (OskarStark)

This PR was merged into the 2.8 branch.

Discussion
----------

added missing "

Commits
-------

1baf05a added missing "

fabpot committed:
feature #14904 [toolbar] Merged colored icons in toolbar (MJBGO)

This PR was squashed before being merged into the 2.8 branch (closes #14904).

Discussion
----------

[toolbar] Merged colored icons in toolbar

| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets |
| License       | MIT
| Doc PR        |

Status icons are doubled without any apparent reason.

Token before
![capture_token_before](https://cloud.githubusercontent.com/assets/8344487/8023946/203c6fc8-0d22-11e5-9d8f-2760bfdf0c18.PNG)

Token after
![capture_token_after](https://cloud.githubusercontent.com/assets/8344487/8024004/d0f4e9b6-0d23-11e5-9263-dd2a38b14d12.PNG)

Security before
![capture_security_before](https://cloud.githubusercontent.com/assets/8344487/8023949/2441d6ee-0d22-11e5-9fed-168d850633d2.PNG)

Security after
![capture_security_after](https://cloud.githubusercontent.com/assets/8344487/8023950/2444a1da-0d22-11e5-9c8a-1deb654e7ac4.PNG)

Commits
-------

5b67bbd [toolbar] Merged colored icons in toolbar

fabpot committed:
feature #14903 [profiler][request toolbar] Removed route name from the toolbar (MJBGO)

This PR was merged into the 2.8 branch.

Discussion
----------

[profiler][request toolbar] Removed route name from the toolbar

| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #14744
| License       | MIT
| Doc PR        |

The action block takes nearly 1/3 of the bar width which is not good.
The controller + action names are useful but the route name less important.

Before
![capture_before](https://cloud.githubusercontent.com/assets/8344487/8023606/325acb00-0d13-11e5-86db-ab2a08242227.PNG)

After
![capture_after](https://cloud.githubusercontent.com/assets/8344487/8023580/908edc22-0d11-11e5-92bd-37b380cc3433.PNG)

The route name is shown by hovering the action block
![capture_tooltip](https://cloud.githubusercontent.com/assets/8344487/8023584/a5ae08f8-0d11-11e5-97a5-ffec6d3e41cc.PNG)

Commits
-------

9004e29 Removed route name from the debugbar.

jakzal committed:
[HttpFoundation] Postpone setting the date header on a Response

fabpot committed:
Merge branch '2.7' into 2.8

* 2.7:
  [Console] SymfonyStyle : Fix blocks wordwrapping
  [Console] SymfonyStyle : Fix blocks output is broken on windows cmd
  [Validator] remove partial deprecation annotation
  Updated UPGRADE-2.4.md
  [Form] Support DateTimeImmutable in transform()
  Show the FormType and FormTypeExtension in case of deprecated use of setDefaultOptions
  [FrameworkBundle] Document form.csrf_provider service deprecation
  [Form] add test to avoid regression of #14891
  without this change allways the legacy code get called
  [Form] Fix call to removed method (BC broken in 2.3)
  Fix ask and askHidden methods
  [HttpFoundation] Get response content as resource several times for PHP >= 5.6
  Change error message to reflect SecurityContext deprecation.
  fixed merge
  Issue #14815
  [Console] SymfonyStyle : fix & automate block gaps.
  [Console] SymfonyStyle : Improve EOL consistency by relying on output instance
  Improved duplicated code in FileLocator

Tobion committed:
feature #14912 [HttpFoundation] Postpone setting the date header on a Response (jakzal)

This PR was merged into the 2.8 branch.

Discussion
----------

[HttpFoundation] Postpone setting the date header on a Response

| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #14906
| License       | MIT
| Doc PR        | -

The only risk of doing this is if someone called `getDate()` and the date header was not present, the date might be slightly different than the one sent with the headers.

`getDate()` could also set the date header first time it's requested (do a lazy initialisation).

Commits
-------

2ad3b0d [HttpFoundation] Postpone setting the date header on a Response

fabpot committed:
Merge branch '2.7' into 2.8

* 2.7:
  Fix test name
  fixed CS
  Allow new lines in Messages translated with transchoice() (replacement for #14867)
  [Form] Swap new ChoiceView constructor arguments to ease migrating from the deprecated one
  [2.3] Fix tests on Windows
  [Yaml] remove partial deprecation annotation
  Silence invasive deprecation warnings, opt-in for warnings
  Documenting how to keep option value BC - see #14377

Conflicts:
	src/Symfony/Bridge/Doctrine/composer.json
	src/Symfony/Bridge/Twig/composer.json

gimler committed:
fix Merge branch '2.7' into 2.8 JsonFileLoader

fabpot committed:
minor #14954 fix Merge branch '2.7' into 2.8 JsonFileLoader (gimler)

This PR was merged into the 2.8 branch.

Discussion
----------

fix Merge branch '2.7' into 2.8 JsonFileLoader

This fix a merge commit see https://github.com/symfony/symfony/commit/5593bdd56e5146693ac0437edb0175d897c2f9bd. The `stream_is_local` and `file_exists` checks are all ready done in the parent `FileLoader` class.

| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | no
| Fixed tickets | none
| License       | MIT
| Doc PR        | none

Commits
-------

692deff fix Merge branch '2.7' into 2.8 JsonFileLoader

aitboudad committed:
[Profiler][Translation] added filter.

fabpot committed:
feature #14993 [Profiler][Translation] added filter. (aitboudad)

This PR was merged into the 2.8 branch.

Discussion
----------

[Profiler][Translation] added filter.

| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Fixed tickets  | ~
| Tests pass?   | yes
| License       | MIT

![selection_017](https://cloud.githubusercontent.com/assets/1753742/8159914/c2c8ad68-135a-11e5-9ae3-f2e055ea3230.jpg)

Commits
-------

65f9291 [Profiler][Translation] added filter.

Commit:
[HttpKernel] Added env & debug to Bundle dependencies method

andrerom commented:
Closed and reopened against 2.8 *with env and debug arguments* as #15011
